### PR TITLE
Feature/performance improvements

### PR DESCRIPTION
**Pull request by fivetran-catfritz**
## PR Overview
**This PR will address the following Issue/Feature:**
- #100 
- #39

**This PR will result in the following new package version:**

- v0.13.0 breaking for materialization changes

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

> ## 🚨 Breaking Changes 🚨
> > ⚠️ Since the following changes are breaking, a `--full-refresh` after upgrading will be required.
> - Performance improvements:
>   - Snowflake, Postgres, and Redshift:
>     - Added an incremental strategy for the following models:
>       - `int_netsuite2__tran_with_converted_amounts`
>       - `netsuite2__balance_sheet`
>       - `netsuite2__income_statement`
>       - `netsuite2_transaction_details`
>   - Bigquery and Databricks
>     - Due to the variation in pricing and runtime priorities for customer, by default we chose table instead of incremental materialization for Bigquery and Databricks. For more information on this decision, see the [Incremental Strategy section](https://github.com/fivetran/dbt_netsuite/blob/main/DECISIONLOG.md#incremental-strategy) of the DECISIONLOG.
>     - To enable incremental materialization for these destinations, see the [Incremental Materialization section](https://github.com/fivetran/dbt_netsuite/blob/main/README.md#-adding-incremental-materialization-for-bigquery-and-databricks) of the README for instructions.
> 
> ## Features
> - Added a default 3-day look-back to incremental models to accommodate late arriving records, based on the `_fivetran_synced_date` of transaction records. The number of days can be changed by setting the var `lookback_window` in your dbt_project.yml. See the [Lookback Window section of the README](https://github.com/fivetran/dbt_netsuite/blob/main/README.md#lookback-window) for more details. 

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [x] dbt run (if incremental models are present)

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [ ] ~docs were regenerated (unless this PR does not include any code or yml updates)~
  - will regenerate docs in the release branch to avoid merge conflicts
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

- see internal ticket

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -232,6 +232,36 @@ vars:
...
+    netsuite2:
+      netsuite2__income_statement:
+        +materialized: incremental # default is table for Bigquery and Databricks
+      netsuite2__transaction_details:
+        +materialized: incremental # default is table for Bigquery and Databricks
+      netsuite2__balance_sheet:
+        +materialized: incremental # default is table for Bigquery and Databricks
+      intermediate:
+        int_netsuite2__tran_with_converted_amounts:
+          +materialized: incremental # default is ephemeral for Bigquery and Databricks
```
Comment (fivetran-joemarkiewicz) [inline README.md:262]: When running the models with this setup I get the following error. Any thoughts as to what this error is a result of?

image

Comment (fivetran-joemarkiewicz) [reply README.md:262]: Then if I only change the materialization of the end models (ignoring the intermediate model) I get the following error.

image

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-catfritz thanks for making these changes, I have a few comments and suggestions below. Let me know if you would like to discuss any of my notes in more detail. Thanks!

File: README.md (hunk 1 above)
Comment (fivetran-catfritz) [reply README.md:262]: What variables and/or data were you using? If your selections produce duplicates, then the merge won't work. The variables I put in my validation should allow this to work, but let me know if it still doesn't. @fivetran-joemarkiewicz

Comment (fivetran-joemarkiewicz) [reply README.md:262]: @fivetran-catfritz ahhh okay let me try your variables. Is this something you feel we need to callout directly?

File: models/netsuite2/netsuite2__balance_sheet.sql [hunk 2]
```diff
@@ -129,132 +153,133 @@ balance_sheet as (
 
     {% endif %}
 
-  from transactions_with_converted_amounts
-  
-  --Below is only used if balance sheet transaction detail columns are specified dbt_project.yml file.
```
Comment (jmongerlyra) [inline models/netsuite2/netsuite2__balance_sheet.sql:134]: @fivetran-catfritz Will we still have the ability to pass through columns from transaction detail? I may have missed it somewhere.

**Comment by jmongerlyra** [comment]
@fivetran-catfritz I'm getting the errors below when running the new models. We are using Snowflake. Any suggestions?

`dbt run --select +netsuite2__balance_sheet +netsuite2__income_statement +netsuite2__transaction_details`

```
14:42:34    Database Error in model netsuite2__balance_sheet (models/netsuite2/netsuite2__balance_sheet.sql)
  002036 (42601): SQL compilation error:
  Subquery containing correlated aggregate function [MAX(INT_NETSUITE2__TRAN_WITH_CONVERTED_AMOUNTS._FIVETRAN_SYNCED_DATE)] can only appear in having or select clause
14:42:34
14:42:34    Database Error in model netsuite2__income_statement (models/netsuite2/netsuite2__income_statement.sql)
  002036 (42601): SQL compilation error:
  Subquery containing correlated aggregate function [MAX(INT_NETSUITE2__TRAN_WITH_CONVERTED_AMOUNTS._FIVETRAN_SYNCED_DATE)] can only appear in having or select clause
14:42:34
14:42:34    Database Error in model netsuite2__transaction_details (models/netsuite2/netsuite2__transaction_details.sql)
  002036 (42601): SQL compilation error:
  Subquery containing correlated aggregate function [MAX(INT_NETSUITE2__TRAN_WITH_CONVERTED_AMOUNTS._FIVETRAN_SYNCED_DATE)] can only appear in having or select clause
```

**Comment by fivetran-catfritz** [comment]
@jmongerlyra Hi Jared thanks for trying out the update and sharing the error! I think I found a workaround for your error and pushed it to my test branch. Does installing the update resolve your error?

**Comment by rwang-lyra** [comment]
> _FIVETRAN_SYNCED_DATE

I wonder if this a new field not yet released by fivetran, hence the error

**Comment by fivetran-catfritz** [comment]
@jmongerlyra I have one more suggestion, and then yes a call might be beneficial. `_FIVETRAN_SYNCED_DATE` is a field I added to the source package, so you might need to do a full refresh on the the source models before running the transforms.

**Comment by jmongerlyra** [comment]
@fivetran-catfritz I'm using the command below with `+` in front of the models. I think that should build all dependencies, but I am not a dbt expert. 

`dbt run --select +netsuite2__balance_sheet +netsuite2__income_statement +netsuite2__transaction_details`

I am pulling from here in `packages.yml`. I think it's in turn pulling the correct branch from `dbt_netsuite_source`. If a meeting would be helpful, I think you have my email. If not, I can send you something.

```
  - git: https://github.com/fivetran/dbt_netsuite.git
    revision: feature/performance-improvements
    warn-unpinned: false
```

**Comment by fivetran-catfritz** [comment]
@jmongerlyra Ahh the first time you try these new changes, you'll need to throw a `--full-refresh` at the end to rebuild the upstreams completely. I'm not sure how Snowflake is handling all the materialization updates, but that usually is what is needed. It would look like:
`dbt run --select +netsuite2__balance_sheet +netsuite2__income_statement +netsuite2__transaction_details --full-refresh`.

Then, subsequent runs you would want to run without the `--full-refresh` to test out the incremental strategy. 

If that doesn't work, then yes let's have a meeting. You can schedule some time on my calendar via [this link](https://calendar.app.google/LRM8nDT1rczs3RYT9). Thanks!

File: README.md [hunk 3]
```diff
@@ -232,6 +232,36 @@ vars:
 ### Override the data models variable
 This package is designed to run **either** the Netsuite.com or Netsuite2 data models. However, for documentation purposes, an additional variable `netsuite_data_model_override` was created to allow for both data model types to be run at the same time by setting the variable value to `netsuite`. This is only to ensure the [dbt docs](https://fivetran.github.io/dbt_netsuite/) (which is hosted on this repository) is generated for both model types. While this variable is provided, we recommend you do not adjust the variable and instead change the `netsuite_data_model` variable to fit your configuration needs.
 
+### Lookback Window
+Records from the source can sometimes arrive late. Since several of the models in this package are incremental, by default we look back 3 days to ensure late arrivals are captured while avoiding the need for frequent full refreshes. While the frequency can be reduced, we still recommend running `dbt --full-refresh` periodically to maintain data quality of the models.
```
Comment (jmongerlyra) [inline README.md:236]: @fivetran-catfritz Can you elaborate on the dates being compared? If it's the `last_modified_date` seems like this value could be small. If it's the effective date of the transaction, prior period modifications could be an issue.

**Comment by jmongerlyra** [comment]
@fivetran-catfritz I ran with `--full-refresh` and the data looks good. 👍 The balance sheet and income statement models still match our environment. We have the features below enabled.

```
    netsuite2__multibook_accounting_enabled: true
    netsuite2__using_to_subsidiary: true
```

**Comment by jmongerlyra** [comment]
Thanks for including us!

>     After your first successful run, were you able to do a 2nd run without the full refresh? I ask because the new incremental logic will only kick in when you don't use a full refresh.

Yes, I was able to perform a second run, removing the `--full-refresh` flag.

>     If so, would you say the models run faster on the 2nd/subsequent runs?

Much faster. The build normally takes 20 minutes. The incremental run completed in 3 minutes.

>     From searching around, my understanding is that shorter runtimes means a lower Snowflake cost. Would you agree with this?

Agree that this change should result in lower Snowflake costs depending on customer's individual agreements.

**Comment by fivetran-catfritz** [comment]
Thanks @jmongerlyra for your input, pleased to hear the changes were working as I hoped. We will be proceeding with releasing these updates, so I post back here when they are released!

File: README.md (hunk 1 above)
Comment (fivetran-catfritz) [reply README.md:262]: 2nd update--after our discussion and deep dive, changing the default behavior is not necessary as we were just seeing artifacts from our own data. Reverting default behavior back to the original false.

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-catfritz great work on these update! @jmongerlyra and @rwang-lyra thanks for chiming in on this thread and providing feedback and confirmation of the performance updates and data quality prior to our release! Your collaboration is much extremely valuable!

@fivetran-catfritz with the latest updates, this looks good to me to be merged and ready for release!

**Pull request merged**